indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java

```diff
@@ -880,55 +888,79 @@ private void removeWorker(final Worker worker)
     lazyWorkers.remove(worker.getHost());
   }
 
+  /**
+   * Schedule a task that will, at some point in the future, clean up znodes and issue failures for "tasksToFail"
+   * if they are being run by "worker".
+   */
   private void scheduleTasksCleanupForWorker(final String worker, final List<String> tasksToFail)
   {
-    removedWorkerCleanups.put(
-        worker, cleanupExec.schedule(
-            new Runnable()
-            {
-              @Override
-              public void run()
-              {
-                log.info("Running scheduled cleanup for Worker[%s]", worker);
-                try {
-                  for (String assignedTask : tasksToFail) {
-                    String taskPath = JOINER.join(indexerZkConfig.getTasksPath(), worker, assignedTask);
-                    String statusPath = JOINER.join(indexerZkConfig.getStatusPath(), worker, assignedTask);
-                    if (cf.checkExists().forPath(taskPath) != null) {
-                      cf.delete().guaranteed().forPath(taskPath);
-                    }
+    // This method is only called from the PathChildrenCache event handler, so this may look like a race,
+    // but is actually not.
+    cancelWorkerCleanup(worker);
 
-                    if (cf.checkExists().forPath(statusPath) != null) {
-                      cf.delete().guaranteed().forPath(statusPath);
-                    }
-
-                    log.info("Failing task[%s]", assignedTask);
-                    RemoteTaskRunnerWorkItem taskRunnerWorkItem = runningTasks.remove(assignedTask);
-                    if (taskRunnerWorkItem != null) {
-                      taskRunnerWorkItem.setResult(TaskStatus.failure(taskRunnerWorkItem.getTaskId()));
-                    } else {
-                      log.warn("RemoteTaskRunner has no knowledge of task[%s]", assignedTask);
-                    }
-                  }
-
-                  // worker is gone, remove worker task status announcements path.
-                  String workerStatusPath = JOINER.join(indexerZkConfig.getStatusPath(), worker);
-                  if (cf.checkExists().forPath(workerStatusPath) != null) {
-                    cf.delete().guaranteed().forPath(JOINER.join(indexerZkConfig.getStatusPath(), worker));
-                  }
+    final ListenableScheduledFuture<?> cleanupTask = cleanupExec.schedule(
+        new Runnable()
+        {
+          @Override
+          public void run()
+          {
+            log.info("Running scheduled cleanup for Worker[%s]", worker);
+            try {
+              for (String assignedTask : tasksToFail) {
+                String taskPath = JOINER.join(indexerZkConfig.getTasksPath(), worker, assignedTask);
+                String statusPath = JOINER.join(indexerZkConfig.getStatusPath(), worker, assignedTask);
+                if (cf.checkExists().forPath(taskPath) != null) {
+                  cf.delete().guaranteed().forPath(taskPath);
```

Comment:
Same question about a race between this statement and the immediately prior line.